account/payments.go

```diff
@@ -340,69 +618,106 @@ func (a *Service) getMaxAmount(destination string, routeHints []*lnrpc.RouteHint
 		a.log.Errorf("lnclient.ListChannels error: %v", err)
 		return 0, fmt.Errorf("lnclient.ListChannels error: %w", err)
 	}
+
 	var totalMax uint64
+OUTER:
 	for _, c := range channels.Channels {
-		if c.LocalBalance == 0 {
+		// Calculate how much can be sent through this channel after accounting for channel reserve
+		thisChannelCanSendSat, minHtlcMsat := channelConstraints(c)
+
+		if thisChannelCanSendSat*1000 <= int64(minHtlcMsat) {
 			continue
 		}
-		a.log.Infof("cid: %v, active: %v, balance: %v, channnel reserve: %v", c.ChanId, c.Active, c.LocalBalance, c.LocalConstraints.ChanReserveSat)
-		_, err := lnclient.QueryRoutes(context.Background(), &lnrpc.QueryRoutesRequest{
+
+		a.log.Infof("cid: %v, active: %v, balance: %v, can send: %v, channel reserve: %v, minHtlcMsat: %v", c.ChanId, c.Active, c.LocalBalance, thisChannelCanSendSat, c.LocalConstraints.ChanReserveSat, minHtlcMsat)
+
+		// Try to find a route for the maximum amount we can send through this channel
+		// Start with a small probe amount to check if the route exists
+		routes, err := lnclient.QueryRoutes(context.Background(), &lnrpc.QueryRoutesRequest{
```

Comment:
In general we don't really need to query route for every channel as we can reuse the same route we get in the first time and only the first policy changes.
The down side that it will complicate the implementation so if you think the time to query the route has a small affect than I am OK to keep it as is.

Reply:
I've tried this with 2 channels and the entire function took 50 milliseconds. That seems reasonable enough.